hedera-node/hedera-app/src/main/java/com/hedera/node/app/history/impl/HistoryLibraryImpl.java

```diff
@@ -120,6 +111,18 @@ public Bytes proveChainOfTrust(
     @Override
     public boolean verifyChainOfTrust(@NonNull final Bytes proof) {
         requireNonNull(proof);
-        return BRIDGE.verifyChainOfTrust(SNARK_KEYS.verifyingKey(), proof.toByteArray());
+        return BRIDGE.verifyChainOfTrust(snarkKeys().verifyingKey(), proof.toByteArray());
+    }
+
+    private ProvingAndVerifyingSnarkKeys snarkKeys() {
+        try {
+            if (SNARK_KEYS == null) {
+                final var elf = HistoryLibraryBridge.loadAddressBookRotationProgram();
+                SNARK_KEYS = BRIDGE.snarkVerificationKey(elf);
+            }
```

Comment:
Is it safe that this could be initialized multiple times for concurrent requests? If not, could add double checked locking.

Reply:
As far as I know it will be called from `handleThread` , so should not be an issue. But I can change `SNARK_KEYS` to volatile. Thanks @steven-sheehy